core-bundle/src/Command/MigrateCommand.php

```diff
@@ -0,0 +1,278 @@
+            return 1;
+        }
+
+        if (!$this->executeMigrations()) {
+            return 1;
+        }
+
+        $this->io->success('All migrations completed.');
+
+        return 0;
+    }
+
+    private function executeMigrations(): bool
+    {
+        while (true) {
+            $first = true;
+
+            foreach ($this->migrations->getPendingMigrations() as $migration) {
+                if ($first) {
+                    $this->io->section('Pending migrations');
+                    $first = false;
+                }
+
+                $this->io->writeln(' * '.$migration);
+            }
+
+            $runOnceFiles = $this->getRunOnceFiles();
+
+            if ($runOnceFiles) {
+                @trigger_error('Using runonce files has been deprecated and will no longer work in Contao 5.0. Use the migration framework instead.', E_USER_DEPRECATED);
+            }
+
+            foreach ($runOnceFiles as $file) {
+                if ($first) {
+                    $this->io->section('Pending migrations');
+                    $first = false;
+                }
+
+                $this->io->writeln(' * Runonce file: '.$file);
+            }
+
+            if ($first) {
+                break;
+            }
+
+            if (!$this->io->confirm('Execute the listed migrations?')) {
+                return false;
+            }
+
+            $this->io->section('Execute migrations');
+
+            foreach ($this->migrations->runMigrations() as $result) {
+                $this->io->writeln(' * '.$result->getMessage());
+
+                if (!$result->isSuccessful()) {
+                    $this->io->error('Migration failed');
+                }
+            }
+
+            foreach ($this->getRunOnceFiles() as $file) {
+                $this->executeRunonceFile($file);
+                $this->io->writeln(' * Executed runonce file: '.$file);
+            }
+
+            $this->io->success('Executed migrations.');
+        }
+
+        return true;
+    }
+
+    private function getRunOnceFiles(): array
+    {
+        try {
+            $files = $this->fileLocator->locate('config/runonce.php', null, false);
+        } catch (FileLocatorFileNotFoundException $e) {
+            return [];
+        }
+
+        return array_map(function ($path) {
+            return rtrim((new Filesystem())->makePathRelative($path, $this->projectDir), '/');
+        }, $files);
+    }
+
+    private function executeRunonceFile(string $file): void
+    {
+        $this->framework->initialize();
+
+        include $this->projectDir.'/'.$file;
+
+        (new Filesystem())->remove($this->projectDir.'/'.$file);
+    }
+
+    private function executeSchemaDiff(bool $completeOption): bool
+    {
+        if (null === $this->installer) {
+            $this->io->error('Service contao.installer of contao/installation-bundle not found. The installation bundle needs to be installed in order to execute schema diff migrations.');
```

Comment:
are we planning to get away from the installer? Asking because the schema diff can be done through Doctrine, but as far as I remember the Installer might add additional stuff (like collation migration?).

Reply:
AFAIK we are planning to get away from the installer, but this does not need to be part of this PR I think.